### PR TITLE
Converted p3 microphysics cloud_rain_accretion subroutine to C++/Kokkos.

### PR DESCRIPTION
**Pull request by jeff-cohere**
I've done the conversion for the `cloud_rain_accretion` function, and I'm submitting the work as is, since I'm not sure about the details of merging PRs on E3SM these days.

It looks like I need to merge some stuff from the master branch and then squash some commits, but it might help to see the individual commits before I do that.

### CODE REVIEW

File: components/scream/src/physics/p3/tests/p3_cloud_rain_acc_unit_tests.cpp [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+  };
+
+  // Run the Fortran subroutine.
+  for (Int i = 0; i < num_runs; ++i) {
+    cloud_rain_accretion(fortran_data[i]);
+  }
+
+  // Run the C++ subroutine.
+  for (Int i = 0; i < num_runs; ++i) {
+    CloudRainAccretionData& d = cxx_data[i];
```
Comment (jgfouca) [inline components/scream/src/physics/p3/tests/p3_cloud_rain_acc_unit_tests.cpp]: Unfortunately, this pattern won't work. The _f functions convert the incoming scalars into uniform packs as a hack. For the unit tests, we want to test non-uniform packs. See the pattern in the bfb unit tests in p3_ice_collection_unit_tests.cpp

File: components/scream/src/physics/p3/CMakeLists.txt [hunk 2]
```diff
@@ -31,9 +31,10 @@ if (NOT CUDA_BUILD)
     p3_functions_dsd2.cpp
     p3_functions_find.cpp
     p3_functions_math.cpp
-    p3_functions_autoconversion.cpp
+    p3_functions_autoconversion.cpp)
```
[the remark this answers is not in the document]
Comment (jeff-cohere) [reply components/scream/src/physics/p3/CMakeLists.txt]: Whoops! Wasn't careful enough with that merge. Thanks.

File: components/scream/src/physics/p3/tests/p3_cloud_rain_acc_unit_tests.cpp (hunk 1 above)
Comment (jeff-cohere) [reply components/scream/src/physics/p3/tests/p3_cloud_rain_acc_unit_tests.cpp]: Cool, thanks for the pointer.

Comment (jeff-cohere) [reply components/scream/src/physics/p3/tests/p3_cloud_rain_acc_unit_tests.cpp]: Should I emulate the pack size logic in the ice collection unit test as well? I can pull in the device sync stuff too. I copied this from another unit test that didn't have those things, so I can use this test as a model for the others.

Comment (jgfouca) [reply components/scream/src/physics/p3/tests/p3_cloud_rain_acc_unit_tests.cpp]: Yes, that test follows the correct model for unit-tests of functions that just take single packs of data.

**Comment by jeff-cohere** [comment]
Well, we had some [test failures on Cori](https://gist.github.com/jeff-cohere/f496018899c18eb8c709a907e2e893fa), but none related to this (relatively simple) one.

Let me know if you want to stash that test failure log anywhere, @jgfouca

**Comment by jeff-cohere** [comment]
Thanks Jim! Anything else need to happen before I merge? Do we follow a special procedure or just click the Merge pull request button these days?

**Comment by ambrad** [comment]
Nice!

We have to wait for the Autotester to mark this PR as passed. It's possible the failure on Cori's GPU will show up as a test failure on White, also a GPU machine.

File: components/scream/src/physics/p3/p3_functions_cloud_rain_acc_impl.hpp [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+                       const Spack& qc_incld, const Spack& nc_incld,
+                       const Spack& qr_incld, Spack& qcacc, Spack& ncacc)
+{
+  constexpr Scalar qsmall = C::QSMALL;
+
+  const auto qr_and_qc_not_small = (qr_incld >= qsmall) && (qc_incld >= qsmall);
+  if (qr_and_qc_not_small.any()) {
+    constexpr int iparam = P3C::iparam;
+
+    if (iparam == 1) {
```
Comment (pressel) [inline components/scream/src/physics/p3/p3_functions_cloud_rain_acc_impl.hpp]: We have decided to only use option iparam=3, so the other options don't need to be included.

**Comment by PeterCaldwell** [comment]
Yes, it's true that we're only intending to support iparam=3 at this time. The logic was that we wanted the code to be as easy to port and easy to understand as possible, and that people could add new options in later as they saw fit. This logic only works if the coders know that's the plan though... Sorry Jeff!

**Comment by pressel** [comment]
That's true @jeff-cohere, there was no way for you to know this.  @PeterCaldwell, I really should have removed all of the other iparam options during the P3 F90 refactor. To try to avoid this coming up again in the future, I'll make a note of this on the Micro P3 conversion to C++ task table on confluence.

**Comment by jeff-cohere** [comment]
No problem, gents. This is exactly what code reviews are for! For what it's worth, I didn't spend a lot of time hunting down `iparam` and actually converting the subroutine. It's certainly easy enough to remove the unneeded code.

**Comment by jeff-cohere** [comment]
What do we do with codecov reports like the above (with red !s)? Is this just another automated emission we ignore, or was I supposed to upload a coverage report from the baseline?

**Comment by jgfouca** [comment]
@jeff-cohere , I have not historically paid much attention to the coverage report. As long as each PR has unit tests for any significant addition, I'm happy.

**Comment by jeff-cohere** [comment]
Looks like there was a failure on white. Are the full logs available somewhere?

**Comment by jgfouca** [comment]
@jeff-cohere , you should be able to repeat the error on any GPU machine. The AT baseline comparison compares baselines generated with origin/master with (your PR merged with origin/master).

**Comment by jgfouca** [comment]
You can use the test-all-scream script (with the -i flag) to mimic.

**Comment by ambrad** [comment]
@jgfouca @jeff-cohere I see this routine has a pow in it. I believe that explains the DP and SP failures on White.. Use bfb_pow to get BFB in the BFB builds. Search in micro_p3.F90 for other uses of bfb_pow to see how it's used.

**Comment by jeff-cohere** [comment]
Thanks, @ambrad and @jgfouca . I'll change `pow` to `bfb_pow`. That failure makes sense.

**Comment by jeff-cohere** [comment]
I'm testing the fix on Cori, and then I'll re-request a review here to appease the machine.

**Comment by jeff-cohere** [comment]
Hrm. Cori is behaving oddly at the moment (`make` is hanging). Anyone mind if I send this through again so we can see what happens?

**Comment by ambrad** [comment]
@jeff-cohere I believe we just add the AT:RETEST label to request an AT rerun w/in the next ~4 hours. I added it just now to get that going.

**Comment by jeff-cohere** [comment]
Good to know. Thanks!

**Comment by bartgol** [comment]
> @jeff-cohere I believe we just add the AT:RETEST label to request an AT rerun w/in the next ~4 hours. I added it just now to get that going.

The PR is not going to be tested until the review is complete. This is because the PR is from a _fork_.

The AT can start testing _before_ the reviewers approve if both these are met:

1) the pr is *not* from a fork
2) the user is in a team that is whitelisted

@jeff-cohere I think you are in the 'scream team' user team. So next time, push the branch on the main repo, and issue the PR from there. This way, while you wait for the review, the AT can start testing your branch.

**Comment by jeff-cohere** [comment]
Thanks @bartgol . I appreciate everyone filling me in on the details.

**Comment by bartgol** [comment]
Also, noticed that the two reviews are 'stale', since there was a commit pushed after the reviewers approved. I believe the AT deems the PR 'not inspected' in this case, even if the reviewers approved. I'm going to double check this with the AT developers, but I'm pretty sure that's the case.

**Comment by jeff-cohere** [comment]
Hmm. Investigating the failure on Cori (which seems to be better now).

**Comment by ambrad** [comment]
I marked this PR as Non-BFB because of the bfb_pow change to F90.

**Comment by bartgol** [comment]
I manually tested the PR on gpu against a modified master (where 'modified' means master plus the commit that added the bfb_pow call). Everything passed, so this PR can be merged in.

@jeff-cohere merge away as soon as you're confident things are fine on cori too.

**Pull request merged**